File: scrape_gsoc.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"sync"
	"time"
)

const (
	baseURL   = "https://summerofcode.withgoogle.com"
	userAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) GSoC-Go-Scraper/1.0"
)

type OrganizationAPI struct {
	Name        string `json:"name"`
	Slug        string `json:"slug"`
	WebsiteURL  string `json:"website_url"`
	IdeasLink   string `json:"ideas_link"`
	SourceCode  string `json:"source_code"`
	Description string `json:"description"`
}

type ProjectAPI struct {
	Title            string `json:"title"`
	BodyShort        string `json:"body_short"`
	Body             string `json:"body"`
	Size             string `json:"size"`
	Status           string `json:"status"`
	ContributorName  string `json:"contributor_name"`
	ProjectCodeURL   string `json:"project_code_url"`
	OrganizationSlug string `json:"organization_slug"`
}

type ProjectsEnvelope struct {
	Entities struct {
		Projects []ProjectAPI `json:"projects"`
	} `json:"entities"`
}

type ProjectSummary struct {
	Title            string `json:"title"`
	Overview         string `json:"overview"`
	Size             string `json:"size"`
	Status           string `json:"status"`
	ContributorName  string `json:"contributor_name"`
	ProjectCodeURL   string `json:"project_code_url"`
	OrganizationSlug string `json:"organization_slug"`
}

type OrgResult struct {
	Name                    string           `json:"name"`
	Slug                    string           `json:"slug"`
	WebsiteURL              string           `json:"website_url"`
	IdeasLink               string           `json:"ideas_link"`
	SourceCode              string           `json:"source_code"`
	OrgOverview             string           `json:"org_overview"`
	GitHubURLsFromIdeaList  []string         `json:"github_urls_from_idea_list_page"`
	ProjectsOverview        []ProjectSummary `json:"projects_overview"`
	ProjectCount            int              `json:"project_count"`
	IdeasScrapeError        *string          `json:"ideas_scrape_error"`
}

type OutputPayload struct {
	Program            string      `json:"program"`
	SourcePage         string      `json:"source_page"`
	OrganizationsCount int         `json:"organizations_count"`
	Organizations      []OrgResult `json:"organizations"`
}

var githubURLRegex = regexp.MustCompile(`https?://(?:www\.)?github\.com/[^\s"'<>]+`)

func fetchJSON(client *http.Client, u string, out any) error {
	req, err := http.NewRequest(http.MethodGet, u, nil)
	if err != nil {
		return err
	}
	req.Header.Set("User-Agent", userAgent)
	req.Header.Set("Accept", "application/json")

	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("request failed %s: status %d", u, resp.StatusCode)
	}

	dec := json.NewDecoder(resp.Body)
	if err := dec.Decode(out); err != nil {
		return err
	}
	return nil
}

func fetchText(client *http.Client, u string) (string, error) {
	req, err := http.NewRequest(http.MethodGet, u, nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("User-Agent", userAgent)

	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return "", fmt.Errorf("request failed %s: status %d", u, resp.StatusCode)
	}

	b, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

func normalizeURL(u string) string {
	return strings.TrimRight(strings.TrimSpace(u), ").,;:'\"")
}

func extractGitHubURLs(text string) []string {
	matches := githubURLRegex.FindAllString(text, -1)
	seen := make(map[string]struct{})
	out := make([]string, 0, len(matches))
	for _, m := range matches {
		clean := normalizeURL(m)
		k := strings.ToLower(clean)
		if _, ok := seen[k]; ok {
			continue
		}
		seen[k] = struct{}{}
		out = append(out, clean)
	}
	return out
}

func orgProjects(client *http.Client, programYear, orgSlug string) ([]ProjectAPI, error) {
	q := url.Values{}
	q.Set("role", "")
	q.Set("program_slug", programYear)
	q.Set("organization_slug", orgSlug)
	u := fmt.Sprintf("%s/api/projects/?%s", baseURL, q.Encode())

	var env ProjectsEnvelope
	if err := fetchJSON(client, u, &env); err != nil {
		return nil, err
	}
	return env.Entities.Projects, nil
}

func summarizeProjects(projects []ProjectAPI) []ProjectSummary {
	out := make([]ProjectSummary, 0, len(projects))
	for _, p := range projects {
		overview := p.BodyShort
		if overview == "" {
			overview = p.Body
		}
		out = append(out, ProjectSummary{
			Title:            p.Title,
			Overview:         overview,
			Size:             p.Size,
			Status:           p.Status,
			ContributorName:  p.ContributorName,
			ProjectCodeURL:   p.ProjectCodeURL,
			OrganizationSlug: p.OrganizationSlug,
		})
	}
	return out
}

func scrapeOneOrg(client *http.Client, programYear string, org OrganizationAPI) OrgResult {
	var githubFromIdeas []string
	var scrapeErr *string

	if org.IdeasLink != "" {
		ideasHTML, err := fetchText(client, org.IdeasLink)
		if err != nil {
			s := err.Error()
			scrapeErr = &s
		} else {
			githubFromIdeas = extractGitHubURLs(ideasHTML)
		}
	}

	projects, err := orgProjects(client, programYear, org.Slug)
	if err != nil {
		empty := []ProjectSummary{}
		s := err.Error()
		if scrapeErr == nil {
			scrapeErr = &s
		}
		return OrgResult{
			Name:                   org.Name,
			Slug:                   org.Slug,
			WebsiteURL:             org.WebsiteURL,
			IdeasLink:              org.IdeasLink,
			SourceCode:             org.SourceCode,
			OrgOverview:            org.Description,
			GitHubURLsFromIdeaList: githubFromIdeas,
			ProjectsOverview:       empty,
			ProjectCount:           0,
			IdeasScrapeError:       scrapeErr,
		}
	}

	projectSummaries := summarizeProjects(projects)
	return OrgResult{
		Name:                   org.Name,
		Slug:                   org.Slug,
		WebsiteURL:             org.WebsiteURL,
		IdeasLink:              org.IdeasLink,
		SourceCode:             org.SourceCode,
		OrgOverview:            org.Description,
		GitHubURLsFromIdeaList: githubFromIdeas,
		ProjectsOverview:       projectSummaries,
		ProjectCount:           len(projectSummaries),
		IdeasScrapeError:       scrapeErr,
	}
}

func main() {
	program := flag.String("program", "2025", "Program year slug")
	outputDir := flag.String("output-dir", "output", "Directory for JSON output")
	workers := flag.Int("workers", 10, "Parallel workers for org scraping")
	flag.Parse()

	if *workers < 1 {
		*workers = 1
	}

	if err := os.MkdirAll(*outputDir, 0o755); err != nil {
		fmt.Fprintf(os.Stderr, "failed to create output dir: %v\n", err)
		os.Exit(1)
	}

	client := &http.Client{Timeout: 25 * time.Second}

	orgsURL := fmt.Sprintf("%s/api/program/%s/organizations/", baseURL, *program)
	var organizations []OrganizationAPI
	if err := fetchJSON(client, orgsURL, &organizations); err != nil {
		fmt.Fprintf(os.Stderr, "failed to fetch organizations: %v\n", err)
		os.Exit(1)
	}

	jobs := make(chan OrganizationAPI)
	resultsCh := make(chan OrgResult)
	var wg sync.WaitGroup

	for i := 0; i < *workers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for org := range jobs {
				resultsCh <- scrapeOneOrg(client, *program, org)
			}
		}()
	}

	go func() {
		for _, org := range organizations {
			jobs <- org
		}
		close(jobs)
		wg.Wait()
		close(resultsCh)
	}()

	results := make([]OrgResult, 0, len(organizations))
	for r := range resultsCh {
		results = append(results, r)
	}

	sort.Slice(results, func(i, j int) bool {
		return strings.ToLower(results[i].Name) < strings.ToLower(results[j].Name)
	})

	payload := OutputPayload{
		Program:            *program,
		SourcePage:         fmt.Sprintf("%s/programs/%s/organizations", baseURL, *program),
		OrganizationsCount: len(results),
		Organizations:      results,
	}

	outFile := filepath.Join(*outputDir, fmt.Sprintf("gsoc_%s_organizations_projects.json", *program))
	data, err := json.MarshalIndent(payload, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "failed to encode output json: %v\n", err)
		os.Exit(1)
	}

	if err := os.WriteFile(outFile, data, 0o644); err != nil {
		fmt.Fprintf(os.Stderr, "failed to write output file: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("Wrote: %s\n", outFile)
}

File: scrape_gsoc.py
#!/usr/bin/env python3
import argparse
import json
import re
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path
from typing import Any
from urllib.parse import urlencode
from urllib.request import Request, urlopen

BASE_URL = "https://summerofcode.withgoogle.com"
UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) GSoC-Scraper/1.0"


def fetch_json(url: str, timeout: int = 20) -> Any:
    req = Request(url, headers={"User-Agent": UA, "Accept": "application/json"})
    with urlopen(req, timeout=timeout) as resp:
        return json.loads(resp.read().decode("utf-8"))


def fetch_text(url: str, timeout: int = 20) -> str:
    req = Request(url, headers={"User-Agent": UA})
    with urlopen(req, timeout=timeout) as resp:
        return resp.read().decode("utf-8", errors="replace")


def norm_url(url: str) -> str:
    return url.rstrip(").,;:'\"")


def extract_github_urls(text: str) -> list[str]:
    matches = re.findall(r"https?://(?:www\.)?github\.com/[^\s\"'<>]+", text, flags=re.IGNORECASE)
    deduped = []
    seen = set()
    for raw in matches:
        cleaned = norm_url(raw)
        key = cleaned.lower()
        if key not in seen:
            seen.add(key)
            deduped.append(cleaned)
    return deduped


def org_projects(program_year: str, organization_slug: str) -> list[dict[str, Any]]:
    query = urlencode(
        {
            "role": "",
            "program_slug": program_year,
            "organization_slug": organization_slug,
        }
    )
    url = f"{BASE_URL}/api/projects/?{query}"
    payload = fetch_json(url)
    entities = payload.get("entities", {})
    projects = entities.get("projects", [])
    if isinstance(projects, list):
        return projects
    return []


def summarize_projects(projects: list[dict[str, Any]]) -> list[dict[str, Any]]:
    summarized: list[dict[str, Any]] = []
    for p in projects:
        summarized.append(
            {
                "title": p.get("title"),
                "overview": p.get("body_short") or p.get("body"),
                "size": p.get("size"),
                "status": p.get("status"),
                "contributor_name": p.get("contributor_name"),
                "project_code_url": p.get("project_code_url"),
                "organization_slug": p.get("organization_slug"),
            }
        )
    return summarized


def scrape_one_org(program_year: str, org: dict[str, Any]) -> dict[str, Any]:
    slug = org.get("slug")
    ideas_link = org.get("ideas_link")
    source_code = org.get("source_code")

    github_from_ideas: list[str] = []
    ideas_scrape_error = None
    if ideas_link:
        try:
            ideas_html = fetch_text(ideas_link)
            github_from_ideas = extract_github_urls(ideas_html)
        except Exception as exc:  # narrow surface: report per-org failure, continue overall scrape
            ideas_scrape_error = str(exc)

    projects = org_projects(program_year, slug)

    return {
        "name": org.get("name"),
        "slug": slug,
        "website_url": org.get("website_url"),
        "ideas_link": ideas_link,
        "source_code": source_code,
        "org_overview": org.get("description"),
        "github_urls_from_idea_list_page": github_from_ideas,
        "projects_overview": summarize_projects(projects),
        "project_count": len(projects),
        "ideas_scrape_error": ideas_scrape_error,
    }


def main() -> None:
    parser = argparse.ArgumentParser(description="Scrape GSoC organizations, GitHub links, and project overviews.")
    parser.add_argument("--program", default="2025", help="Program year slug (default: 2025)")
    parser.add_argument("--output-dir", default="output", help="Directory for JSON output")
    parser.add_argument("--workers", type=int, default=10, help="Parallel workers for org scraping")
    args = parser.parse_args()

    output_dir = Path(args.output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    orgs_url = f"{BASE_URL}/api/program/{args.program}/organizations/"
    organizations: list[dict[str, Any]] = fetch_json(orgs_url)

    items: list[dict[str, Any]] = []
    with ThreadPoolExecutor(max_workers=max(1, args.workers)) as pool:
        futures = [pool.submit(scrape_one_org, args.program, org) for org in organizations]
        for fut in as_completed(futures):
            items.append(fut.result())

    items.sort(key=lambda x: (x.get("name") or "").lower())

    payload = {
        "program": args.program,
        "source_page": f"{BASE_URL}/programs/{args.program}/organizations",
        "organizations_count": len(items),
        "organizations": items,
    }

    out_file = output_dir / f"gsoc_{args.program}_organizations_projects.json"
    out_file.write_text(json.dumps(payload, indent=2, ensure_ascii=False), encoding="utf-8")
    print(f"Wrote: {out_file.resolve()}")


if __name__ == "__main__":
    main()
